include/execinfo.h

```diff
@@ -27,15 +27,43 @@
 
 #include <nuttx/sched.h>
 
+#include <assert.h>
+#include <stdio.h>
 #include <sys/types.h>
 
 /****************************************************************************
  * Pre-processor Definitions
  ****************************************************************************/
 
-#define backtrace(buffer, size) sched_backtrace(_SCHED_GETTID(), \
-                                                buffer, size, 0)
-#define dump_stack()            sched_dumpstack(_SCHED_GETTID())
+/* 3: ' 0x' prefix */
+
+#define BACKTRACE_PTR_FMT_WIDTH  ((int)sizeof(uintptr_t) * 2 + 3)
+
+/* Buffer size needed to hold formatted `depth` backtraces */
+
+#define BACKTRACE_BUFFER_SIZE(d) (BACKTRACE_PTR_FMT_WIDTH * (d) + 1)
+
+#define backtrace(b, s) sched_backtrace(_SCHED_GETTID(), b, s, 0)
+#define dump_stack()    sched_dumpstack(_SCHED_GETTID())
+
+/* Format a backtrace into a buffer for dumping. */
+
+#define backtrace_format(buf, buflen, backtrace, depth)              \
+do                                                                   \
+  {                                                                  \
+    FAR const char *format = "%0*p ";                                \
+    int i;                                                           \
+                                                                     \
+    DEBUGASSERT(buflen >= BACKTRACE_BUFFER_SIZE(depth));             \
+    buf[0] = '\0';                                                   \
+    for (i = 0; i < depth && backtrace[i]; i++)                      \
+      {                                                              \
```

Comment:
check the buflen before format

Reply:
already assert at line 57

Reply:
Common function should allow partial formatting, rather than assertions

Reply:
Done